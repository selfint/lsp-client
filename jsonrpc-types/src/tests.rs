use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
pub(crate) struct Params {
    pub p0: u32,
    pub p1: u32,
}

#[cfg(test)]
macro_rules! snapshot {
    ($e:expr) => {
        insta::assert_json_snapshot!($e);

        let serialized = serde_json::to_value($e);
        let deserialized = serde_json::from_value(serialized.unwrap()).unwrap();

        similar_asserts::assert_eq!($e, deserialized);
    };
}

#[cfg(test)]
pub(crate) use snapshot;
