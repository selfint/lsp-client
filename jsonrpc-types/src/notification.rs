use serde::{ser::SerializeStruct, Deserialize, Serialize};

use crate::JSONRPC_V2;

#[derive(Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
pub struct Notification<P> {
    pub method: String,
    pub params: P,
}

impl<P> Notification<P> {
    pub fn new(method: impl Into<String>, params: P) -> Self {
        Self {
            method: method.into(),
            params,
        }
    }
}

impl<P: Serialize> Serialize for Notification<P> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut state = serializer.serialize_struct("Notification", 3)?;
        state.serialize_field("jsonrpc", JSONRPC_V2)?;
        state.serialize_field("method", &self.method)?;
        state.serialize_field("params", &self.params)?;
        state.end()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::tests::{snapshot, Params};

    #[test]
    fn test_notification_serde() {
        snapshot!(Notification::new("method", ()));
        snapshot!(Notification::new("method", vec![0, 1]));
        snapshot!(Notification::new("method", Params { p0: 0, p1: 1 }));
    }
}
