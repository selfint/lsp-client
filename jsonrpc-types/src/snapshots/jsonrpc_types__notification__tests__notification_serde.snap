---
source: jsonrpc-types/src/notification.rs
expression: "Notification::new(\"method\", ())"
---
{
  "jsonrpc": "2.0",
  "method": "method",
  "params": null
}
